tstest/integration: data race

### What is the issue?

Failed during CI:

https://github.com/tailscale/tailscale/runs/4173401996?check_suite_focus=true

```
==================
WARNING: DATA RACE
Write at 0x00c0007110b0 by goroutine 646:
  runtime.closechan()
      /opt/hostedtoolcache/go/1.17.3/x64/src/runtime/chan.go:355 +0x0
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMap()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:630 +0xa6b
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMachine()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:262 +0x34a
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMachine-fm()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:239 +0x57
  net/http.HandlerFunc.ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2046 +0x4d
  net/http.(*ServeMux).ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2424 +0xc5
  tailscale.com/tstest/integration/testcontrol.(*Server).ServeHTTP()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:204 +0x8f
  net/http.serverHandler.ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2878 +0x89a
  net/http.(*conn).serve()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:1929 +0x12e4
  net/http.(*Server).Serve·dwrap·87()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:3033 +0x58

Previous read at 0x00c0007110b0 by goroutine 8:
  runtime.chansend()
      /opt/hostedtoolcache/go/1.17.3/x64/src/runtime/chan.go:158 +0x0
  tailscale.com/tstest/integration/testcontrol.sendUpdate()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:541 +0x45
  tailscale.com/tstest/integration/testcontrol.(*Server).SetExpireAllNodes()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:164 +0x119
  tailscale.com/tstest/integration.TestOneNodeExpiredKey()
      /home/runner/work/tailscale/tailscale/tstest/integration/integration_test.go:114 +0x3e9
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1306 +0x47

Goroutine 646 (running) created at:
  net/http.(*Server).Serve()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:3033 +0x847
  net/http/httptest.(*Server).goServe.func1()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/httptest/server.go:308 +0xc4

Goroutine 8 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1504 +0x9d1
  tailscale.com/tstest/integration.TestMain()
      /home/runner/work/tailscale/tailscale/tstest/integration/integration_test.go:54 +0xe8
  main.main()
      _testmain.go:63 +0x264
==================
==================
WARNING: DATA RACE
Write at 0x00c0006e36e0 by goroutine 646:
  runtime.mapassign_fast64()
      /opt/hostedtoolcache/go/1.17.3/x64/src/runtime/map_fast64.go:92 +0x0
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMap()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:635 +0xb10
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMachine()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:262 +0x34a
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMachine-fm()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:239 +0x57
  net/http.HandlerFunc.ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2046 +0x4d
  net/http.(*ServeMux).ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2424 +0xc5
  tailscale.com/tstest/integration/testcontrol.(*Server).ServeHTTP()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:204 +0x8f
  net/http.serverHandler.ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2878 +0x89a
  net/http.(*conn).serve()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:1929 +0x12e4
  net/http.(*Server).Serve·dwrap·87()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:3033 +0x58

Previous read at 0x00c0006e36e0 by goroutine 8:
  runtime.mapaccess1_fast64()
      /opt/hostedtoolcache/go/1.17.3/x64/src/runtime/map_fast64.go:12 +0x0
  tailscale.com/tstest/integration/testcontrol.(*Server).SetExpireAllNodes()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:164 +0xfd
  tailscale.com/tstest/integration.TestOneNodeExpiredKey()
      /home/runner/work/tailscale/tailscale/tstest/integration/integration_test.go:114 +0x3e9
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1306 +0x47

Goroutine 646 (running) created at:
  net/http.(*Server).Serve()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:3033 +0x847
  net/http/httptest.(*Server).goServe.func1()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/httptest/server.go:308 +0xc4

Goroutine 8 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1504 +0x9d1
  tailscale.com/tstest/integration.TestMain()
      /home/runner/work/tailscale/tailscale/tstest/integration/integration_test.go:54 +0xe8
  main.main()
      _testmain.go:63 +0x264
==================
==================
WARNING: DATA RACE
Write at 0x00c000555b48 by goroutine 646:
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMap()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:635 +0xb24
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMachine()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:262 +0x34a
  tailscale.com/tstest/integration/testcontrol.(*Server).serveMachine-fm()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:239 +0x57
  net/http.HandlerFunc.ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2046 +0x4d
  net/http.(*ServeMux).ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2424 +0xc5
  tailscale.com/tstest/integration/testcontrol.(*Server).ServeHTTP()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:204 +0x8f
  net/http.serverHandler.ServeHTTP()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:2878 +0x89a
  net/http.(*conn).serve()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:1929 +0x12e4
  net/http.(*Server).Serve·dwrap·87()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:3033 +0x58

Previous read at 0x00c000555b48 by goroutine 8:
  tailscale.com/tstest/integration/testcontrol.(*Server).SetExpireAllNodes()
      /home/runner/work/tailscale/tailscale/tstest/integration/testcontrol/testcontrol.go:164 +0x107
  tailscale.com/tstest/integration.TestOneNodeExpiredKey()
      /home/runner/work/tailscale/tailscale/tstest/integration/integration_test.go:114 +0x3e9
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1306 +0x47

Goroutine 646 (running) created at:
  net/http.(*Server).Serve()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/server.go:3033 +0x847
  net/http/httptest.(*Server).goServe.func1()
      /opt/hostedtoolcache/go/1.17.3/x64/src/net/http/httptest/server.go:308 +0xc4

Goroutine 8 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.17.3/x64/src/testing/testing.go:1504 +0x9d1
  tailscale.com/tstest/integration.TestMain()
      /home/runner/work/tailscale/tailscale/tstest/integration/integration_test.go:54 +0xe8
  main.main()
      _testmain.go:63 +0x264
==================
--- FAIL: TestOneNodeExpiredKey (3.66s)
    integration.go:74: built [tailscale.com/cmd/tailscaled tailscale.com/cmd/tailscale] in 1.842s
    integration_test.go:494: DERP httpsrv listener: 127.0.0.1:42383
    integration_test.go:98: Status: NeedsLogin
    integration_test.go:739: Running [up --login-server=http://127.0.0.1:39001] ...
    stuntest.go:64: STUN server shutdown
    testing.go:1152: race detected during execution of test
--- FAIL: TestCollectPanic (3.01s)
    integration.go:74: built [tailscale.com/cmd/tailscaled tailscale.com/cmd/tailscale] in 1.899s
    integration_test.go:494: DERP httpsrv listener: 127.0.0.1:44689
    integration_test.go:143: initial run: 2021/11/11 02:25:49 You have enabled a non-default log target. Doing without being told to by Tailscale staff or your network administrator will make getting support difficult.
        logtail started
        Program starting: v1.17.0-date.20211101, Go 1.17.3-race: []string{"/tmp/TestCollectPanic1681370557/001/tailscaled", "--cleanup"}
        LogID: 4a89c6d5d0964856a4b2264d312f609b2a82ed942888b42a261413b612c958af
        logpolicy: using system state directory "/home/runner/.local/share/tailscale"
        logpolicy.Read /home/runner/.local/share/tailscale/tailscaled.log.conf: open /home/runner/.local/share/tailscale/tailscaled.log.conf: no such file or directory
        flushing log.
        logger closing down
        logtail: dialed "127.0.0.1:37067" in 1ms
    stuntest.go:64: STUN server shutdown
    testing.go:1152: race detected during execution of test
FAIL
exit status 1
FAIL	tailscale.com/tstest/integration	19.336s
```

### Steps to reproduce

_No response_

### Are there any recent changes that introduced the issue?

_No response_

### OS

Linux

### OS version

_No response_

### Tailscale version

_No response_

### Bug report

_No response_

--- a/tstest/integration/testcontrol/testcontrol.go
+++ b/tstest/integration/testcontrol/testcontrol.go
@@ -157,10 +157,11 @@ func (s *Server) AddPingRequest(nodeKeyDst key.NodePublic, pr *tailcfg.PingReque
 // Mark the Node key of every node as expired
 func (s *Server) SetExpireAllNodes(expired bool) {
 	s.mu.Lock()
-	s.allExpired = expired
-	s.mu.Unlock()
+	defer s.mu.Unlock()
 
-	for _, node := range s.AllNodes() {
+	s.allExpired = expired
+
+	for _, node := range s.nodes {
 		sendUpdate(s.updates[node.ID], updateSelfChanged)
 	}
 }

